### PR TITLE
cmd/deps: add args to filter output

### PR DESCRIPTION
**Pull request by cho-m**
* `--direct-build` to only show direct build dependencies, which gives information on what formulae are needed when `--build-from-source`.
* `--direct-test` to only show direct test dependencies when outputting in `--tree` or `--graph` format
* `--skip` to prune parts of dependency tree
* `--level` to limit recursion depth. For now, only added to `--tree` as the support there is simple and idea for arg is similar to what the `tree` & `eza` formula provide.

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Example:
```console
❯ brew deps --include-build libxml2 --tree --annotate
libxml2
├── python-setuptools  [build]
│   └── python@3.12  [build] [test]
│       ├── pkg-config  [build]
│       ├── mpdecimal
│       ├── openssl@3
│       │   └── ca-certificates
│       ├── sqlite
│       │   └── readline
│       └── xz
├── python@3.11  [build] [test]
│   ├── pkg-config  [build]
│   ├── mpdecimal
│   ├── openssl@3
│   │   └── ca-certificates
│   ├── sqlite
│   │   └── readline
│   └── xz
├── python@3.12  [build] [test]
│   ├── pkg-config  [build]
│   ├── mpdecimal
│   ├── openssl@3
│   │   └── ca-certificates
│   ├── sqlite
│   │   └── readline
│   └── xz
├── icu4c
└── readline

❯ brew deps --include-build libxml2 --tree --annotate --direct-build
libxml2
├── python-setuptools  [build]
├── python@3.11  [build] [test]
│   ├── mpdecimal
│   ├── openssl@3
│   │   └── ca-certificates
│   ├── sqlite
│   │   └── readline
│   └── xz
├── python@3.12  [build] [test]
│   ├── mpdecimal
│   ├── openssl@3
│   │   └── ca-certificates
│   ├── sqlite
│   │   └── readline
│   └── xz
├── icu4c
└── readline

❯ brew deps --include-build libxml2 --tree --annotate --direct-build --skip readline,python@3.11
libxml2
├── python-setuptools  [build]
├── python@3.12  [build] [test]
│   ├── mpdecimal
│   ├── openssl@3
│   │   └── ca-certificates
│   ├── sqlite
│   └── xz
└── icu4c

❯ brew deps --include-build libxml2 --tree --annotate --direct-build --skip readline,python@3.11 --level 2
libxml2
├── python-setuptools  [build]
├── python@3.12  [build] [test]
│   ├── mpdecimal
│   ├── openssl@3
│   ├── sqlite
│   └── xz
└── icu4c

❯ brew deps --include-build libxml2 --tree --annotate --direct-build --skip readline,python@3.11 --level 1
libxml2
├── python-setuptools  [build]
├── python@3.12  [build] [test]
└── icu4c
```

### CODE REVIEW

**Comment by cho-m** [comment]
> The `--direct-test` and `--direct-build` options could be expressed with `--direct --include-build` and `--direct --include-test`, right? 

`--direct --include-build` will only output a single level. e.g. 
```console
❯ brew deps --direct --include-build --tree ruby --annotate
ruby
├── autoconf  [build]
├── pkg-config  [build]
├── rust  [build]
├── libyaml
└── openssl@3
```

This is not what is installed when you run `brew install --build-from-source ruby`. 
```console
❯ brew deps --include-build --tree ruby --annotate --direct-build
ruby
├── autoconf  [build]
│   └── m4
├── pkg-config  [build]
├── rust  [build]
│   ├── libgit2
│   │   ├── libssh2
│   │   │   └── openssl@3
│   │   │       └── ca-certificates
│   │   └── openssl@3
│   │       └── ca-certificates
│   ├── libssh2
│   │   └── openssl@3
│   │       └── ca-certificates
│   ├── llvm@17
│   │   └── zstd
│   │       ├── lz4
│   │       └── xz
│   ├── openssl@3
│   │   └── ca-certificates
│   └── pkg-config
├── libyaml
└── openssl@3
    └── ca-certificates
```

The usage of this is analyzing problems that occur when particular formulae are installed in build environment, which often happens in Homebrew/core.

---

> Maybe `--depth` would make more sense as the option name than `--level`?

I don't mind either way. I went with option I was used to using for directory tree printing, e.g. 
```console
❯ tree --help | rg ' -L'
  -L level      Descend only level directories deep.

❯ eza --help | rg ' -L'
  -L, --level DEPTH          limit the depth of recursion
```

File: Library/Homebrew/cmd/deps.rb [hunk 1]
```diff
@@ -88,6 +102,9 @@ class Deps < AbstractCommand
       def run
         raise UsageError, "`brew deps --os=all` is not supported" if args.os == "all"
         raise UsageError, "`brew deps --arch=all` is not supported" if args.arch == "all"
+        if (level = args.level) && !level.match?(/^[1-9][0-9]*$/)
```
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/deps.rb:105]: Doing some basic testing: it doesn't look like a `rescue` is required as invalid input will just end up being `0` from `to_i`.

**Comment by apainintheneck** [comment]
> The usage of this is analyzing problems that occur when particular formulae are installed in build environment, which often happens in Homebrew/core.
> 
> In most cases, I've fallen back to grepping and scripting as this doesn't seem possible in current `brew` commands without excessive calls to `brew` with extra post processing.

In that case, maybe `brew deps --build-from-source` would make more sense than `brew deps --include-build --direct-build`. I think that this option should conflict with the existing `--include-*` and `--skip-*` options as well since it's not easy to determine precedence here.

File: Library/Homebrew/dependencies_helpers.rb [hunk 2]
```diff
@@ -17,19 +17,28 @@ def args_includes_ignores(args)
...
+    [includes, ignores, recursive_ignores]
   end
 
-  def recursive_includes(klass, root_dependent, includes, ignores)
+  def recursive_includes(klass, root_dependent, includes, ignores, recursive_ignores: nil, skip: nil)
     raise ArgumentError, "Invalid class argument: #{klass}" if klass != Dependency && klass != Requirement
 
     cache_key = "recursive_includes_#{includes}_#{ignores}"
+    cache_key = "#{cache_key}_#{recursive_ignores}" if recursive_ignores.present?
+    cache_key = "#{cache_key}_#{skip}" if skip.present?
```
Comment (apainintheneck) [inline Library/Homebrew/dependencies_helpers.rb:32]: I wonder if it's really still useful to have a cache key that's this complex to create.

**Review by apainintheneck — COMMENTED** [review]
Another option would be to have `--direct` take a list of dependencies types. That would allow you to run `brew deps --include-build --direct=build NAME` to get the build from source dependencies.

**Comment by MikeMcQuaid** [comment]
> In that case, maybe `brew deps --build-from-source` would make more sense than `brew deps --include-build --direct-build`

Yeh, I really like this idea, nice one @apainintheneck 💡

**Pull request closed without merging**